File: Sources/ConnectorControlState/ServerDelta.swift
import ConnectorControlCore

/// What a regenerated Claude config now runs that it did not before, by
/// connector name (catalog §1.8). An mcpServers entry is a command Claude
/// executes, so a notification about an adopted synced change names the
/// entries rather than saying only that "the list changed".
public struct ServerDelta: Equatable, Sendable {
    public var added: [String]
    public var removed: [String]
    public var changed: [String]

    public init(added: [String] = [], removed: [String] = [], changed: [String] = []) {
        self.added = added
        self.removed = removed
        self.changed = changed
    }

    public init(from before: [String: JSONValue], to after: [String: JSONValue]) {
        added = after.keys.filter { before[$0] == nil }.sorted()
        removed = before.keys.filter { after[$0] == nil }.sorted()
        changed = after.keys.filter { before[$0] != nil && before[$0] != after[$0] }.sorted()
    }

    public var isEmpty: Bool { added.isEmpty && removed.isEmpty && changed.isEmpty }

    /// "adds a, b; removes c; changes d" — at most `limit` names per part, then "and N more".
    public func summary(limit: Int = 4) -> String {
        var parts: [String] = []
        if !added.isEmpty { parts.append("adds " + ServerDelta.list(added, limit: limit)) }
        if !removed.isEmpty { parts.append("removes " + ServerDelta.list(removed, limit: limit)) }
        if !changed.isEmpty { parts.append("changes " + ServerDelta.list(changed, limit: limit)) }
        return parts.joined(separator: "; ")
    }

    static func list(_ names: [String], limit: Int) -> String {
        guard names.count > limit else { return names.joined(separator: ", ") }
        return names.prefix(limit).joined(separator: ", ") + " and \(names.count - limit) more"
    }
}

File: windows/src/ConnectorControl.Core/State/ServerDelta.cs
namespace ConnectorControl.Core.State;

/// <summary>
/// What a regenerated Claude config now runs that it did not before, by
/// connector name (Swift <c>ServerDelta</c>). An mcpServers
/// entry is a command Claude executes, so a notification about an adopted
/// synced change names the entries rather than saying only that "the list changed".
/// </summary>
public sealed record ServerDelta(IReadOnlyList<string> Added, IReadOnlyList<string> Removed, IReadOnlyList<string> Changed)
{
    public static ServerDelta Between(IReadOnlyDictionary<string, JsonValue> before, IReadOnlyDictionary<string, JsonValue> after)
    {
        var added = after.Keys.Where(k => !before.ContainsKey(k)).Order(StringComparer.Ordinal).ToList();
        var removed = before.Keys.Where(k => !after.ContainsKey(k)).Order(StringComparer.Ordinal).ToList();
        var changed = after.Keys.Where(k => before.TryGetValue(k, out var was) && !was.Equals(after[k])).Order(StringComparer.Ordinal).ToList();
        return new ServerDelta(added, removed, changed);
    }

    public bool IsEmpty => Added.Count == 0 && Removed.Count == 0 && Changed.Count == 0;

    /// <summary>"adds a, b; removes c; changes d" — at most <paramref name="limit"/> names per part, then "and N more".</summary>
    public string Summary(int limit = 4)
    {
        var parts = new List<string>();
        if (Added.Count > 0)
        {
            parts.Add("adds " + List(Added, limit));
        }
        if (Removed.Count > 0)
        {
            parts.Add("removes " + List(Removed, limit));
        }
        if (Changed.Count > 0)
        {
            parts.Add("changes " + List(Changed, limit));
        }
        return string.Join("; ", parts);
    }

    private static string List(IReadOnlyList<string> names, int limit) =>
        names.Count <= limit
            ? string.Join(", ", names)
            : string.Join(", ", names.Take(limit)) + $" and {names.Count - limit} more";
}
